doc/dev_guide/c_style_guide_new.rst

```diff
@@ -0,0 +1,920 @@
+some random integer loop counter, it should probably be called ``i``.
+Calling it ``loop_counter`` is non-productive, if there is no chance of it
+being mis-understood. Similarly, ``tmp`` can be just about any type of
+variable that is used to hold a temporary value.
+
+If you are afraid to mix up your local variable names, you have another
+problem, which is called the function-growth-hormone-imbalance syndrome.
+See chapter 6 (Functions).
+
+For function naming we have a convention is to use:
+
+*    ``new``/``delete`` for functions which
+     allocate + initialize and destroy + deallocate an object,
+*    ``create``/``destroy`` for functions which initialize/destroy an object
+     but do not handle memory management,
+*    ``init``/``free`` for functions which initialize/destroy libraries and subsystems.
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Chapter 5: Typedefs
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Please don't use things like ``vps_t``.
+It's a **mistake** to use typedef for structures and pointers. When you see a
+
+..  code-block:: c
+
+    vps_t a;
+
+in the source, what does it mean?
+In contrast, if it says
+
+..  code-block:: c
+
+    struct virtual_container *a;
+
+you can actually tell what ``a`` is.
+
+Lots of people think that typedefs ``help readability``. Not so. They are
+useful only for:
+
+#.  totally opaque objects (where the typedef is actively used to **hide**
+    what the object is).
+
+    Example: ``pte_t`` etc. opaque objects that you can only access using
+    the proper accessor functions.
+
+    ..  note::
+
+        Opaqueness and ``accessor functions`` are not good in themselves.
+        The reason we have them for things like pte_t etc. is that there
+        really is absolutely **zero** portably accessible information there.
+
+#.  Clear integer types, where the abstraction **helps** avoid confusion
+    whether it is ``int`` or ``long``.
+
+    u8/u16/u32 are perfectly fine typedefs, although they fit into
+    point 4 better than here.
+
+    ..  note::
+
+        Again - there needs to be a **reason** for this. If something is
+        ``unsigned long``, then there's no reason to do
+        typedef unsigned long myflags_t;
+
+    but if there is a clear reason for why it under certain circumstances
+    might be an ``unsigned int`` and under other configurations might be
+    ``unsigned long``, then by all means go ahead and use a typedef.
+
+#.  when you use sparse to literally create a **new** type for
+    type-checking.
+
+#.  New types which are identical to standard C99 types, in certain
+    exceptional circumstances.
+
+    Although it would only take a short amount of time for the eyes and
+    brain to become accustomed to the standard types like ``uint32_t``,
+    some people object to their use anyway.
+
+    Therefore, the Linux-specific ``u8/u16/u32/u64`` types and their
```

Comment:
We don't use these types. They can only be found in old SQL code, but we are getting rid of them.